drs_downloader/cli.py

```diff
@@ -1,11 +1,183 @@
+@click.option("--destination_dir", "-d", show_default=True,
+              default="/tmp/testing", help="Destination directory.")
+@click.option("--manifest_path", "-m", show_default=True,
+              help="Path to manifest tsv.")
+@click.option('--drs_header', default='ga4gh_drs_uri', show_default=True,
+              help='The column header in the TSV file associated with the DRS URIs.'
+                   'Example: pfb:ga4gh_drs_uri')
+@click.option('--api_key_path', show_default=True,
+              help='Gen3 credentials file')
+@click.option('--endpoint', show_default=True, required=True,
+              help='Gen3 endpoint')
+def gen3(silent: bool, destination_dir: str, manifest_path: str, drs_header: str, api_key_path: str, endpoint: str):
+    """Copy files from gen3 server."""
+    # read from manifest
+    ids_from_manifest = _extract_tsv_info(Path(manifest_path), drs_header)
+
+    _perform_downloads(destination_dir,
+                       Gen3DrsClient(api_key_path=api_key_path, endpoint=endpoint),
+                       ids_from_manifest,
+                       silent)
+
+
+def _perform_downloads(destination_dir, drs_client, ids_from_manifest,  silent):
+    """Common helper method to run downloads."""
+    # verify parameters
+    if destination_dir:
+        destination_dir = Path(destination_dir)
+        destination_dir.mkdir(parents=True, exist_ok=True)
+    # create a manager
+    drs_manager = DrsAsyncManager(drs_client=drs_client, show_progress=not silent)
+
+    # call the server, get size, checksums etc.; sort them by size
+    drs_objects = drs_manager.get_objects(ids_from_manifest)
+    drs_objects.sort(key=lambda x: x.size, reverse=False)
+    # optimize based on workload
+    drs_objects = drs_manager.optimize_workload(drs_objects)
+    # determine the total number of batches
+    total_batches = len(drs_objects) / DEFAULT_MAX_SIMULTANEOUS_OBJECT_SIGNERS
+    if math.ceil(total_batches) - total_batches > 0:
+        total_batches += 1
+        total_batches = int(total_batches)
+    for chunk_of_drs_objects in tqdm.tqdm(
+            DrsAsyncManager.chunker(drs_objects, DEFAULT_MAX_SIMULTANEOUS_OBJECT_SIGNERS),
+            total=total_batches,
+            desc="TOTAL_DOWNLOAD_PROGRESS", leave=True):
+
+        drs_manager.download(chunk_of_drs_objects, destination_dir)
+
+    # show results
+    if not silent:
+        for drs_object in drs_objects:
+            if len(drs_object.errors) == 0:
+                logger.info((drs_object.name, 'OK', drs_object.size, len(drs_object.file_parts)))
+        logger.info(('done', 'statistics.max_files_open', drs_client.statistics.max_files_open))
+
+    at_least_one_error = False
+    for drs_object in drs_objects:
+        if len(drs_object.errors) > 0:
+            logger.error((drs_object.name, 'ERROR', drs_object.size, len(drs_object.file_parts), drs_object.errors))
+            at_least_one_error = True
+    if at_least_one_error:
+        exit(99)
```

Comment:
I recommend returning the value `1` in the case of general errors.
Although any positive exit code indicates errors, in my experience, the value `1` is most commonly used.

Additional positive values may be used to indicate specific types of errors, yet IMO that is not needed for this tool.

Reply:
Agreed.  Will address today in a change to this PR, including test.

Reply:
Fixed and [tested](https://github.com/anvilproject/drs_downloader/blob/feature/manager/tests/integration/test_problems.py#L9)